c-bindings-gen/src/main.rs

```diff
@@ -0,0 +1,809 @@
+		println_docs(w, &var.attrs, "\t");
+		if let syn::Fields::Unit = var.fields {} else { unimplemented!(); }
+		if var.discriminant.is_some() { unimplemented!(); }
+		write!(w, "\t{},\n", var.ident).unwrap();
+	}
+	write!(w, "}}\nuse {}::{} as ln{};\nimpl {} {{\n", module_path, e.ident, e.ident, e.ident).unwrap();
+	write!(w, "\t#[allow(unused)]\n\tpub(crate) fn to_ln(&self) -> ln{} {{\n\t\tmatch self {{\n", e.ident).unwrap();
+	for var in e.variants.iter() {
+		write!(w, "\t\t\t{}::{} => ln{}::{},\n", e.ident, var.ident, e.ident, var.ident).unwrap();
+	}
+	write!(w, "\t\t}}\n\t}}\n").unwrap();
+	write!(w, "\t#[allow(unused)]\n\tpub(crate) fn from_ln(lnt: ln{}) -> Self {{\n\t\tmatch lnt {{\n", e.ident).unwrap();
+	for var in e.variants.iter() {
+		write!(w, "\t\t\tln{}::{} => {}::{},\n", e.ident, var.ident, e.ident, var.ident).unwrap();
+	}
+	write!(w, "\t\t}}\n\t}}\n}}\n").unwrap();
+}
+
+fn should_export(path: &str) -> bool {
+	match path {
+		f if f.contains("lib.rs") => true,
+		f if f.contains("mod.rs") => true,
+		f if f.contains("features.rs") => true,
+		f if f.contains("config.rs") => true,
+		f if f.contains("events.rs") => true,
+		f if f.contains("errors.rs") => true,
+		f if f.contains("logger.rs") => true,
+		f if f.contains("peer_handler.rs") => true,
+		f if f.contains("msgs.rs") => true,
+		f if f.contains("chaininterface.rs") => true,
+		f if f.contains("channelmanager.rs") => true,
+		f if f.contains("chan_utils.rs") => true,
+		f if f.contains("channelmonitor.rs") => true,
+		f if f.contains("keysinterface.rs") => true,
+		f if f.contains("transaction.rs") => true,
+		_ => false,
+	}
+}
+
+fn convert_file(path: &str, out_path: &str, orig_crate: &str, module: &str, header_file: &mut File) {
+	if !should_export(path) { return; }
+	eprintln!("Converting {}...", path);
+
+	let mut file = File::open(path).expect("Unable to open file");
+	let mut src = String::new();
+	file.read_to_string(&mut src).expect("Unable to read file");
+	let syntax = syn::parse_file(&src).expect("Unable to parse file");
+
+	assert!(syntax.shebang.is_none()); // Not sure what this is, hope we dont have one
+
+	let mut out = std::fs::OpenOptions::new().write(true).create(true).truncate(true)
+		.open(&out_path).expect("Unable to open new src file");
+
+	assert_eq!(export_status(&syntax.attrs), ExportStatus::Export);
+	println_docs(&mut out, &syntax.attrs, "");
+
+	let mut type_resolver = TypeResolver::new(module);
+	let orig_module = orig_crate.to_string() + "::" + module;
+
+	if path.ends_with("lib.rs") {
+		write!(out, "#![allow(non_camel_case_types)]\n").unwrap();
+		write!(out, "#![allow(non_snake_case)]\n").unwrap();
+		write!(out, "#![allow(unused_imports)]\n").unwrap();
+		write!(out, "#![allow(unused_variables)]\n").unwrap();
+		write!(out, "mod c_types;\n").unwrap();
+		write!(out, "mod bitcoin;\n").unwrap();
+	} else {
+		write!(out, "\nuse std::ffi::c_void;\nuse bitcoin::hashes::Hash;\n\n").unwrap();
```

Comment:
`writeln!` would be a bit more readable.

Reply:
Lol this after meticulously replacing println with write \n...